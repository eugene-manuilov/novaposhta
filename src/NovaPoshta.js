import Model from './Model';
import Schema from './Schema';
import ERRORS from './Errors';

export default (() => {
	const privateProps = new WeakMap();

	class NovaPoshta {

		constructor(options = {}) {
			const self = this;

			self.apiKey = options.apiKey || '';
			self.endpoint = options.endpoint || 'https://api.novaposhta.ua/v2.0/json/';
			self.logger = options.logger || false;

			const models = {};
			Object.keys(Schema).forEach((model) => {
				const key = model.toLowerCase().substring(0, 1) + model.substring(1);
				models[key] = new Model(self, model, Schema[model]);
			});

			privateProps.set(self, models);
		}

		log(message, params = {}, type = 'info') {
			if (this.logger) {
				this.logger.log(type, message, params);
			}
		}

		_request(params) {
			const self = this;
			const args = {
				method: 'POST',
				body: JSON.stringify({ ...params, apiKey: self.apiKey }),
			};

			return new Promise((resolve, reject) => {
				const onError = (errors, data) => {
					self.log(`NovaPoshta::${params.calledMethod} request failed`, { params, errors }, 'error');
					reject(errors, data);
				};

				fetch(self.endpoint, args)
					.then((response) => {
						if (response.ok) {
							response
								.json()
								.then((json) => {
									if (json.success) {
										self.log(`NovaPoshta::${params.calledMethod} request executed`, { params, response: json }, 'info');
										resolve(json, response);
									} else {
										const errors = [];
										json.errors.forEach((message, i) => {
											const code = json.errorCodes[i];
											errors.push({
												code,
												message,
												en: ERRORS[code] || '',
												ru: ERRORS[code] || '',
												uk: ERRORS[code] || '',
											});
										});

										onError(errors, json);
									}
								})
								.catch((error) => {
									onError([{ message: error.toString() }], null);
								});
						} else {
							onError([{ message: response.statusText }], null);
						}
					})
					.catch((error) => {
						onError([{ message: error.toString() }], null);
					});
			});
		}

		get address() {
			return privateProps.get(this).address;
		}

		get common() {
			return privateProps.get(this).common;
		}

		get counterparty() {
			return privateProps.get(this).counterparty;
		}

		get contactPerson() {
			return privateProps.get(this).contactPerson;
		}

		get internetDocument() {
			return privateProps.get(this).internetDocument;
		}

		get scanSheet() {
			return privateProps.get(this).scanSheet;
		}

		get additionalService() {
			return privateProps.get(this).additionalService;
		}

		get additionalServiceGeneral() {
			return privateProps.get(this).additionalServiceGeneral;
		}

	}

	NovaPoshta.version = '1.1.0';

	return NovaPoshta;
})();
