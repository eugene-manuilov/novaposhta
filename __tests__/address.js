import NovaPoshta from '../src/NovaPoshta';

describe('Address model', () => {
	const api = new NovaPoshta();
	const { address } = api;
	const checkSnapshot = json => expect(JSON.stringify(json)).toMatchSnapshot();

	test('searchSettlements', () => address.searchSettlements({ CityName: '01001' }).then(checkSnapshot));
	test.skip('searchSettlementStreets', () => address.searchSettlementStreets().then(checkSnapshot));
	test.skip('update', () => {});
	test.skip('save', () => {});
	test('getAreas', () => address.getAreas({ Limit: 1 }).then(checkSnapshot));
	test('getCities', () => address.getCities({ Ref: "ebc0eda9-93ec-11e3-b441-0050568002cf" }).then(checkSnapshot));
	test('getSettlements', () => address.getSettlements({ Ref: "0e451e40-4b3a-11e4-ab6d-005056801329" }).then(checkSnapshot));
	test('getWarehouses', () => address.getWarehouses({ CityRef: "20982d74-9b6c-11e2-a57a-d4ae527baec3" }).then(checkSnapshot));
	test('getWarehouseTypes', () => address.getWarehouseTypes().then(checkSnapshot));
	test('getStreet', () => address.getStreet({ CityRef: "20982d74-9b6c-11e2-a57a-d4ae527baec3", Limit: 1 }).then(checkSnapshot));
	test.skip('delete', () => {});
});
