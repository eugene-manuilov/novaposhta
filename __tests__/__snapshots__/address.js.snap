// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address model getAreas 1`] = `"{\\"success\\":true,\\"data\\":[{\\"Ref\\":\\"71508128-9b87-11de-822f-000c2965ae0e\\",\\"AreasCenter\\":\\"db5c88b7-391c-11dd-90d9-001a92567626\\",\\"DescriptionRu\\":\\"АРК\\",\\"Description\\":\\"АРК\\"}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":[],\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;

exports[`Address model getCities 1`] = `"{\\"success\\":true,\\"data\\":[{\\"Description\\":\\"Агрономічне\\",\\"DescriptionRu\\":\\"Агрономичное\\",\\"Ref\\":\\"ebc0eda9-93ec-11e3-b441-0050568002cf\\",\\"Delivery1\\":\\"1\\",\\"Delivery2\\":\\"1\\",\\"Delivery3\\":\\"1\\",\\"Delivery4\\":\\"1\\",\\"Delivery5\\":\\"1\\",\\"Delivery6\\":\\"1\\",\\"Delivery7\\":\\"0\\",\\"Area\\":\\"71508129-9b87-11de-822f-000c2965ae0e\\",\\"SettlementType\\":\\"563ced13-f210-11e3-8c4a-0050568002cf\\",\\"IsBranch\\":\\"0\\",\\"PreventEntryNewStreetsUser\\":null,\\"Conglomerates\\":null,\\"CityID\\":\\"890\\",\\"SettlementTypeDescriptionRu\\":\\"село\\",\\"SettlementTypeDescription\\":\\"село\\",\\"SpecialCashCheck\\":1,\\"AreaDescription\\":\\"Вінницька\\",\\"AreaDescriptionRu\\":\\"Винницкая \\"}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":{\\"totalCount\\":1},\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;

exports[`Address model getSettlements 1`] = `"{\\"success\\":true,\\"data\\":[{\\"Ref\\":\\"0e451e40-4b3a-11e4-ab6d-005056801329\\",\\"SettlementType\\":\\"563ced13-f210-11e3-8c4a-0050568002cf\\",\\"Latitude\\":\\"49.605372000000000\\",\\"Longitude\\":\\"34.335613000000000\\",\\"Description\\":\\"Абазівка\\",\\"DescriptionRu\\":\\"Абазовка\\",\\"SettlementTypeDescription\\":\\"село\\",\\"SettlementTypeDescriptionRu\\":\\"село\\",\\"Region\\":\\"e4ade6ea-4b33-11e4-ab6d-005056801329\\",\\"RegionsDescription\\":\\"Полтавський р-н\\",\\"RegionsDescriptionRu\\":\\"Полтавский р-н\\",\\"Area\\":\\"dcaadf02-4b33-11e4-ab6d-005056801329\\",\\"AreaDescription\\":\\"Полтавська область\\",\\"AreaDescriptionRu\\":\\"Полтавская область\\",\\"Index1\\":\\"38715\\",\\"Index2\\":\\"38715\\",\\"IndexCOATSU1\\":\\"5324080101\\",\\"Delivery1\\":\\"1\\",\\"Delivery2\\":\\"0\\",\\"Delivery3\\":\\"1\\",\\"Delivery4\\":\\"0\\",\\"Delivery5\\":\\"1\\",\\"Delivery6\\":\\"0\\",\\"Delivery7\\":\\"0\\",\\"SpecialCashCheck\\":1,\\"Warehouse\\":\\"1\\"}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":{\\"totalCount\\":1},\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;

exports[`Address model getStreet 1`] = `"{\\"success\\":true,\\"data\\":[{\\"Description\\":\\"24-го Серпня\\",\\"Ref\\":\\"709b80ad-68ce-11e4-acce-0050568002cf\\",\\"StreetsTypeRef\\":\\"Street\\",\\"StreetsType\\":\\"вул.\\"}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":{\\"totalCount\\":\\"26\\"},\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;

exports[`Address model getWarehouseTypes 1`] = `"{\\"success\\":true,\\"data\\":[{\\"Ref\\":\\"6f8c7162-4b72-4b0a-88e5-906948c6a92f\\",\\"Description\\":\\"Parcel Shop\\",\\"DescriptionRu\\":\\"Parcel Shop\\"},{\\"Ref\\":\\"841339c7-591a-42e2-8233-7a0a00f0ed6f\\",\\"Description\\":\\"Поштове відділення\\",\\"DescriptionRu\\":\\"Почтовое отделение\\"},{\\"Ref\\":\\"95dc212d-479c-4ffb-a8ab-8c1b9073d0bc\\",\\"Description\\":\\"Поштомат ПриватБанку\\",\\"DescriptionRu\\":\\"Почтомат приват банка\\"},{\\"Ref\\":\\"9a68df70-0267-42a8-bb5c-37f427e36ee4\\",\\"Description\\":\\"Вантажне відділення\\",\\"DescriptionRu\\":\\"Грузовое отделение\\"},{\\"Ref\\":\\"f9316480-5f2d-425d-bc2c-ac7cd29decf0\\",\\"Description\\":\\"Поштомат\\",\\"DescriptionRu\\":\\"Почтомат\\"}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":[],\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;

exports[`Address model getWarehouses 1`] = `"{\\"success\\":true,\\"data\\":[{\\"SiteKey\\":\\"10119\\",\\"Description\\":\\"Відділення №1: вул. М. Грушевського, 3\\",\\"DescriptionRu\\":\\"Отделение №1: ул. М. Грушевского, 3\\",\\"ShortAddress\\":\\"Заболотів, М. Грушевського, 3\\",\\"ShortAddressRu\\":\\"Заболотов, Грушевского, 3\\",\\"Phone\\":\\"380800500609\\",\\"TypeOfWarehouse\\":\\"9a68df70-0267-42a8-bb5c-37f427e36ee4\\",\\"Ref\\":\\"39931b80-e1c2-11e3-8c4a-0050568002cf\\",\\"Number\\":\\"1\\",\\"CityRef\\":\\"20982d74-9b6c-11e2-a57a-d4ae527baec3\\",\\"CityDescription\\":\\"Заболотів (Снятинський р-н)\\",\\"CityDescriptionRu\\":\\"Заболотов (Снятинский р-н)\\",\\"SettlementRef\\":\\"e717af16-4b33-11e4-ab6d-005056801329\\",\\"SettlementDescription\\":\\"Заболотів\\",\\"SettlementAreaDescription\\":\\"Івано-Франківська область\\",\\"SettlementRegionsDescription\\":\\"Снятинський р-н\\",\\"SettlementTypeDescription\\":\\"селище міського типу\\",\\"Longitude\\":\\"25.284086000000000\\",\\"Latitude\\":\\"48.467503000000000\\",\\"PostFinance\\":\\"1\\",\\"BicycleParking\\":\\"0\\",\\"PaymentAccess\\":\\"0\\",\\"POSTerminal\\":\\"1\\",\\"InternationalShipping\\":\\"0\\",\\"SelfServiceWorkplacesCount\\":\\"0\\",\\"TotalMaxWeightAllowed\\":\\"0\\",\\"PlaceMaxWeightAllowed\\":\\"1000\\",\\"Reception\\":{\\"Monday\\":\\"16:45-18:00\\",\\"Tuesday\\":\\"16:45-18:00\\",\\"Wednesday\\":\\"16:45-18:00\\",\\"Thursday\\":\\"16:45-18:00\\",\\"Friday\\":\\"16:45-18:00\\",\\"Saturday\\":\\"16:30-17:00\\",\\"Sunday\\":\\"10:00-15:00\\"},\\"Delivery\\":{\\"Monday\\":\\"08:00-12:00\\",\\"Tuesday\\":\\"08:00-12:00\\",\\"Wednesday\\":\\"08:00-12:00\\",\\"Thursday\\":\\"08:00-12:00\\",\\"Friday\\":\\"08:00-12:00\\",\\"Saturday\\":\\"09:00-12:00\\",\\"Sunday\\":\\"09:00-14:30\\"},\\"Schedule\\":{\\"Monday\\":\\"09:00-19:00\\",\\"Tuesday\\":\\"09:00-19:00\\",\\"Wednesday\\":\\"09:00-19:00\\",\\"Thursday\\":\\"09:00-19:00\\",\\"Friday\\":\\"09:00-19:00\\",\\"Saturday\\":\\"09:00-17:00\\",\\"Sunday\\":\\"11:00-17:00\\"},\\"DistrictCode\\":\\"Ск За\\",\\"WarehouseStatus\\":\\"Working\\",\\"WarehouseStatusDate\\":\\"2013-04-02 00:00:00\\",\\"CategoryOfWarehouse\\":\\"Branch\\",\\"Direct\\":\\"\\"}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":{\\"totalCount\\":1},\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;

exports[`Address model searchSettlements 1`] = `"{\\"success\\":true,\\"data\\":[{\\"TotalCount\\":1,\\"Addresses\\":[{\\"Present\\":\\"м. Київ, Київська обл.\\",\\"Warehouses\\":2314,\\"MainDescription\\":\\"Київ\\",\\"Area\\":\\"Київська\\",\\"Region\\":\\"\\",\\"SettlementTypeCode\\":\\"м.\\",\\"Ref\\":\\"e718a680-4b33-11e4-ab6d-005056801329\\",\\"DeliveryCity\\":\\"8d5a980d-391c-11dd-90d9-001a92567626\\",\\"AddressDeliveryAllowed\\":true,\\"StreetsAvailability\\":true,\\"ParentRegionTypes\\":\\"область\\",\\"ParentRegionCode\\":\\"обл.\\",\\"RegionTypes\\":\\"\\",\\"RegionTypesCode\\":\\"\\"}]}],\\"errors\\":[],\\"warnings\\":[],\\"info\\":[],\\"messageCodes\\":[],\\"errorCodes\\":[],\\"warningCodes\\":[],\\"infoCodes\\":[]}"`;
